### PR TITLE
[Compiler] Compiling a compiled WF leads to compile errors

### PR DESCRIPTION
**Pull request by EngHabu**
# TL;DR
When a workflow has two nodes both of type WorkflowNodes and attempt to execute the same SubWF, the compiler attempts to compile the SubWorkflow twice (opportunity for optimization).. When it does the second time, it fails because validating Start and End nodes doesn't pass due to them not having targets (by design)... the fix is to handle start and end nodes specially...
The ultimate fix is to avoid compiling twice to begin with.

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Tracking Issue
https://github.com/lyft/flyte/issues/633

### CODE REVIEW

**Review by wild-endeavor — APPROVED** [review]
but why doesn't it fail the first time?  Do they have targets the first time?

**Comment by EngHabu** [comment]
The first time they (start and end nodes) don't even exist, we add them during the compilation phase... so the second time, they look like regular nodes we try to validate...

**Comment by kumare3** [comment]
@EngHabu I am ok with this now, but I feel this is very counter-intuitive. We start with the assumption that we dont have start and end nodes, but then we check for them

**Comment by kumare3** [comment]
Why can we not check for the subworkflow id in a map?

**Review by kumare3 — APPROVED** [review]
